[TEP-0142] Add VolumeMounts to StepAction (#7340)
tektoncd/pipeline: docs/stepactions.md

    date | tee $(results.current-date-human-readable.path)
```

### Declaring VolumeMounts

> You can define `VolumeMounts` in `StepAction`, note that the `name` of the `VolumeMount` should be a **single param reference** to a string param, i.e. `$(params.registryConfig)-foo` is invalid.

Reviewer: ```suggestion
You can define `VolumeMounts` in `StepAction`, note that the `name` of the `VolumeMount` MUST be a **single param reference** to a string param ( for example, `$(params.registryConfig)` is valid but `$(params.registryConfig)-foo` is invalid).
```

> You can define `VolumeMounts` in `StepAction`, note that the `name` of the `VolumeMount` MUST be a **single param reference** to a string param ( for example, `$(params.registryConfig)` is valid but `$(params.registryConfig)-foo` or `"unparametrized-name"` is invalid).

Reviewer 2: must the name be a parameter? can it not just be a string?

Reviewer: Yes, we explain it in the TEP. The name must be a parameter so that the Task using this step action with other steps can provide a volume device with a name they choose.

Otherwise the following can happen:

1. Lets say StepAction1 names a volumeMount `foo`
2. StepAction2 names it `bar`

From the Task perspective, it should be the same volumeDevice/workspace that binds to the mount. The Task is forced to provide two separate volume devices for `foo` and `bar` which might lead to issues. 

Parametrizing it allows the Task author to provide the name and take control of the VolumeDevice/Workspace that will bind to this mount.

Author: Should we include the design decision background in the user facing docs?

Reviewer 2: Thanks for the explanation @chitrangpatel!

 
+1 to @Yongxuanzhang's comment, please add a summary of the above rationale in the docs

Author: added

Note that the `securityContext` from `StepAction` will overwrite the `securityContext` from [`TaskRun`](./taskruns.md/#example-of-running-step-containers-as-a-non-root-user).

### Declaring VolumeMounts

> You can define `VolumeMounts` in `StepAction`, note that the `name` of the `VolumeMount` MUST be a **single param reference** to a string param ( for example, `$(params.registryConfig)` is valid but `$(params.registryConfig)-foo` or `"unparametrized-name"` is invalid). This is to avoid the case where `VolumeMount` from 2 different `StepActions` using the same mounted volume but different names. In this case the Task Author doesn't need to provide 2 separate volume devices.

Reviewer 2: ```suggestion
You can define `VolumeMounts` in `StepActions`. The `name` of the `VolumeMount` MUST be a single reference to a string `Parameter`. For example, `$(params.registryConfig)` is valid while `$(params.registryConfig)-foo` and `"unparametrized-name"` are invalid. This is to ensure reusability of `StepActions` such that `Task` authors have control of which `Volumes` they bind to the `VolumeMounts`.
```
